compiler/test/dotty/tools/dotc/vulpix/RunnerOrchestration.scala

```diff
@@ -0,0 +1,81 @@
+package dotty
+package tools
+package dotc
+package vulpix
+
+import java.io.{
+  File => JFile,
+  InputStream, ObjectInputStream,
+  OutputStream, ObjectOutputStream
+}
+
+import scala.concurrent.duration.Duration
+import scala.collection.mutable
+
+trait RunnerOrchestration {
+
+  /** The maximum amount of active runners, which contain a child JVM */
+  val numberOfSlaves: Int
+
+  /** The maximum duration the child process is allowed to consume before
+   *  getting destroyed
+   */
+  val maxDuration: Duration
+
+  /** Destroy and respawn process after each test */
+  val safeMode: Boolean
+
+  /** Running a `Test` class's main method from the specified `dir` */
+  def runMain(dir: JFile): Status = monitor.runMain(dir)
+
+  private[this] val monitor = new RunnerMonitor
+
+  private class RunnerMonitor {
+
+    def runMain(dir: JFile): Status = withRunner(_.runMain(dir))
+
+    private class Runner(private var process: Process) {
+      def kill(): Unit = ???
+      def runMain(dir: JFile): Status = ???
+    }
+
+    private def createProcess: Process = ???
+
+    private[this] val allRunners =
+      List.fill(numberOfSlaves)(new Runner(createProcess))
+
+    private[this] val freeRunners = mutable.Queue(allRunners: _*)
+    private[this] val busyRunners = mutable.Set.empty[Runner]
+
+    private def getRunner(): Runner = synchronized {
+      while (freeRunners.isEmpty) wait()
```

Comment:
This looks correct, but I've forgotten everything I ever knew about Java concurrency, are you sure there's no existing higher-level concurrency thing from https://docs.oracle.com/javase/8/docs/api/?java/util/concurrent/package-summary.html you could use instead of the wait/notify dance ?